File: 9-Palindrome Number/PalindromeNumber.js
/**
 * @param {number} x
 * @return {boolean}
 */

//1. With built-in methods
var isPalindrome = function(x) {
   
   //1. converting to string and slpit
   let newStr = x.toString().split(''); 
   
   //2. reverse the string, join and converting to a Number
   let reverseX = Number(newStr.reverse().join(""));
   
   if(x === reverseX)
      return true;
   else  return false;
};

//2. Without built-in methods
var isPalindrome = function(x) {
   //Negative numbers cannot be palindromes
   if(x < 0)
      return false; 

   let reversed = 0;
   let original = x;

   //1. number is more than 0
   while(x > 0)
   {
      //2. get last digit
      const lastDigit  = x % 10;//1, 2, 1

      //3. reverse the number
      reversed = reversed * 10 + lastDigit;//1, 1*10+2 = 12, 12*10+1 = 121
   
      //get original 
      x = Math.floor(x/10);//12, 1, 0-end
   }

   return reversed === original;
}

isPalindrome(121); // true
isPalindrome(-121); // false
isPalindrome(10); // false

File: 9-Palindrome Number/PalindromeNumber.java
class Solution {
    public boolean isPalindrome(int x) {
        // Negative numbers are not palindromes
        if (x < 0) {
            return false;
        }

        int result = 0;  // Initialize the variable to store the reversed number
        int n = x;  // Make a copy of the original number

        // Loop to reverse the number
        while (n > 0) {
            int lastDigit = n % 10;  // Get the last digit

            // Check for overflow/underflow
            if (result > (Integer.MAX_VALUE - lastDigit) / 10) {
                return false;  // Return false in case of overflow
            }

            result = result * 10 + lastDigit;  // Append the last digit to the result
            n = n / 10;  // Remove the last digit from n
        }

        // Check if the original number is equal to the reversed number
        return x == result;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test cases
        int num1 = 121;
        int num2 = -121;
        int num3 = 10;

        // Check if each number is a palindrome
        System.out.println(num1 + " is palindrome: " + solution.isPalindrome(num1));
        System.out.println(num2 + " is palindrome: " + solution.isPalindrome(num2));
        System.out.println(num3 + " is palindrome: " + solution.isPalindrome(num3));
    }
}
